### DISCUSSION TITLE
Python 踩坑之旅（一）

### DISCUSSION THREAD

**Discussion by eason8479** [1101-GS4719-C/Python-Community-GS4719-C#910] [Show and tell]
# 前言
以下想和大家分享一下我這學期其它課程中遇到的一些Python的坑。
希望能讓讓大家在將來漫（ㄊㄨㄥˋ）漫（ㄎㄨˇ）的程式路上少踩一些坑
（還有就是我想多寫一些東西來刷一些分）

前情提要，簡而言之，在另一堂課裡，我們小組要寫出一個Line bot
看到這邊，請看官不要緊張，以下的部份都很基礎
基本上，只要有認真在上課的，應該都看得懂
（相信大家都有認真在上課吧！）

# 問題
在Line bot 套件裡，有個功能是傳訊息給使用者
簡單來說，等價於Python 裡的print
但是為了方便維護，所以我把要傳的值放在text裡

（模擬狀況，非實際情況）
``` python
from datetime import datetime

now = datetime.now()
time = str(now.strftime("%H:%M:%S"))

status = '正常'
var1 = '123'
value1 = 'other thing'

text = f'裝置1: machine1\n\
    狀態:{status}\n \
    {var1}:{value1}\n\
    刷新(上次更新:{time})'

print(text)
```

看起來很正常吧，這能有什麼坑？
結果一跑就出錯了
好吧！出錯了就改嘛！
結果把{time}拔掉就好了，但加回去就崩掉了

在繼續往下之前，各位看官也不妨一起想想，問題到底在那？

# 解
雖然我說了，傳訊息給使用者等價於Python 裡的print
但它有個小小的限制，就是不能超過60字元

這件事可以在套件的doc、或是error message 裡找到

但解了一個問題又有另一個問題
大家可以一起數一數，明顯text不到60字元吧！

其實這也很好理解
相信大家都有那位非常堅持自己170 or 180 的朋友吧
不過，身高計通常都會堅持另一個數字

所以，讓我們一起來看看python對text 長度的看法
```python
print(len(text))
```
```
66
```

......好了，現在開始有些尷尬了，所以問題到底在那呢？

# 解的解
後來發現問題其實基本到有些可笑
```python
# 注意二三四行都有縮排
text = f'裝置1: machine1\n\
    狀態:{status}\n \
    {var1}:{value1}\n\
    刷新(上次更新:{time})'
```
縮排的部份居然被算進去了

解法如下(抽查一下，有人知道這在講義的第幾頁嗎？)
```　python
text = (f'裝置1: machine1\n'
    f'狀態:{status}\n'
    f'{var1}:{value1}\n'
    f'刷新(上次更新:{time})')
```

# 結論
最後想和大家分享從坑裡面爬出來後的一些想法
- 基本功真的很重要
- 出問題後，去爬爬套件的doc往往有奇效（善用ctrl+F）
- Error message 真的要看（不然我現在應該還是不知道60字元限制）

題外話：中文字其實和字母一樣都是一個字元長度
所以，如果之後有遇到要壓長度的話不妨改用中文
ex：（machine[7字元]　>　機器[2字元]）

**Comment by HuangHsinTzu** [comment]
恭喜你們找到解決方法!
竟然把縮牌也算進去了
而且居然是大家平常很常忽略的套件doc跟error message立了大功
感謝分享 讓我學到一個酷知識!
ps. 在講義的 7-4 頁!

**Comment by evanchou1204** [comment]
堅持自己170 or 180 的朋友跟身高計通常都會堅持另一個數字的比喻好有才，好喜歡，恭喜你解決問題！
